Translate integrating-with-other-libraries (#117)
reactjs/ja.reactjs.org: content/docs/integrating-with-other-libraries.md

---
id: integrating-with-other-libraries

> title: 他のライブラリとのインテグレーション

Reviewer: とりあえずこのままにしようとは思いますが、ちょっと長いためレイアウト上の問題が出た場合は「統合」「結合」などに変えさせていただく場合があるかもしれませんのでよろしくお願いします

Author: 了解です。

permalink: docs/integrating-with-other-libraries.html
---

React はどのような Web アプリケーションにも適用できます。React は他のアプリケーションに組み込むことができ、また少しだけ気を付ければ、React に他のアプリケーションを組み込むこともできます。ここでは一般的なユースケースである [jQuery](https://jquery.com/) と [Backbone](http://backbonejs.org/) を使った例を紹介しますが、同じ考え方はどのような既存のコードにも適用可能です。

## DOM 操作プラグインとのインテグレーション {#integrating-with-dom-manipulation-plugins}

> React は、React の外で DOM に加えられた変更を認識しません。React 自身の内部表現に基づいて更新を決定します。もし同じ DOM ノードが別のライブラリによって操作された場合、React 的にはコンフリクトしてしまい回復する方法がありません。

Reviewer: ```suggestion
React は、React の外で DOM に加えられた変更を認識しません。React は自身の内部表現に基づいて更新を決定します。もし同じ DOM ノードが別のライブラリによって操作された場合、React は混乱してしまい、回復する方法がありません。
```

ここは直訳でも大丈夫だと思います

> とはいえ、React と DOM 操作プラグインを組み合わせることが不可能であるといっているのではありません。それぞれがやっていることを正しく認識する必要があるのです。

Reviewer: ```suggestion
とはいえ、React と DOM 操作プラグインを組み合わせることが不可能、あるいは必ずしも難しいと言っているのではありません。それぞれがやっていることを正しく認識する必要があるのです。
```

necessarily difficult を加えました

> コンフリクトを回避する最も簡単な方法は、React コンポーネントが更新されないようにすることです。これは、空の`<div />`のように、React から更新する理由がない要素をレンダリングすることで実現できます。

Reviewer: ```suggestion
コンフリクトを回避する最も簡単な方法は、React コンポーネントが更新されないようにすることです。これは、空の`<div />`のように、React から更新する理由がない要素をレンダーすることで実現できます。
```

}
```

`componentDidMount` と `componentWillUnmount` の両方の [ライフサイクルメソッド](/docs/react-component.html#the-component-lifecycle) を定義したことに注意してください。多くの jQuery プラグインは DOM にイベントリスナーをアタッチするので、それらを `componentWillUnmount` でデタッチすることが重要です。もしプラグインがクリーンアップの方法を提供していない場合、あなた自身で提供する必要があります。メモリリークを防ぐためにプラグインが登録したイベントリスナを削除することを忘れないでください。

### jQuery Chosen プラグインとのインテグレーション {#integrating-with-jquery-chosen-plugin}

このアイディアの具体的な例として、`<select>` 要素を拡張する [Chosen](https://harvesthq.github.io/chosen/) プラグインの最小のラッパーを書いてみましょう。

>**Note:**
>
>それが可能だからといって、それが React アプリケーションに最適なアプローチであるという意味ではありません。可能であれば React コンポーネントを使用することをお勧めします。React コンポーネントは React アプリケーションで簡単に再利用でき、また多くの場合、それらの動作や外観をより細かくコントロールできます。

まず、Chosen が DOM に対してどういった操作をしているのか確認しましょう。

`<select>` DOM ノードで Chosen を呼ぶと元の DOM ノードから属性を読み取り、インライン CSS で隠してから、`<select>` の直後に独自の視覚表現を持つ別の DOM ノードを追加します。その後に jQuery イベントを発生させて変更を通知します。

> 我々の <Chosen> というラッパー React コンポーネントで作成したい API は以下のようなものであるとしましょう：

Reviewer: ```suggestion
我々の `<Chosen>` というラッパー React コンポーネントで作成したい API は以下のようなものであるとしましょう：
```

}
```

> わかりやすくするために [uncontrolled component](/docs/uncontrolled-components.html) として実装します。

Reviewer: ```suggestion
わかりやすくするために[非制御コンポーネント](/docs/uncontrolled-components.html)として実装します。
```

> まず、`<select>`を`<div>`で囲んで返す `render()` メソッドを使って空のコンポーネントを作成します：

Reviewer: ```suggestion
まず、`<select>` を `<div>` で囲んで返す `render()` メソッドを持った、空のコンポーネントを作成します：
```

}
```

> 余分な `<div>` で `<select>` をラップしていることに注目してください。これが必要なのは、Chosen は渡された `<select>` の直後に別の DOM 要素を追加するからです。しかし、React からみれば、この `<div>` は常に 1 つの子要素しか持っていません。これにより React による更新が Chosen によって追加された DOM ノードと確実に競合しないようにできるのです。React フローの外側で DOM を変更する場合は、React がその DOM ノードに触れるないようにすることをを確認する必要があります。

Reviewer: ```suggestion
余分な `<div>` で `<select>` をラップしていることに注目してください。これが必要なのは、Chosen は渡された `<select>` の直後に別の DOM 要素を追加するからです。しかし、React からみれば、この `<div>` は常に 1 つの子要素しか持っていません。これにより React による更新が Chosen によって追加された DOM ノードと確実に競合しないようにできるのです。重要なことは、React フローの外側で DOM を変更する場合は、React がその DOM ノードに触る理由を確実になくす必要がある、ということです。
```

[**Try it on CodePen**](http://codepen.io/gaearon/pen/qmqeQx?editors=0010)

React からすると `this.el` フィールドに特別な意味はありません。以前に `render()` メソッドの中で `ref` からこれに代入したことによって成り立っています：

```js
<select className="Chosen-select" ref={el => this.el = el}>
```

> コンポーネントをレンダリングするにはこれで十分ですが、値の変更についても通知を受ける必要があります。この通知を実現するために、Chosen が管理する `<select>` で起こる jQuery の `change` イベントを受け取るようにします。

Reviewer: ```suggestion
コンポーネントをレンダーするにはこれで十分ですが、値の変更について通知を受ける必要もあります。この通知を実現するために、Chosen が管理する `<select>` で起こる jQuery の `change` イベントを受け取るようにします。
```

[**Try it on CodePen**](http://codepen.io/gaearon/pen/xdgKOz?editors=0010)

## 他のビューライブラリとのインテグレーション {#integrating-with-other-view-libraries}

React は [`ReactDOM.render()`](/docs/react-dom.html#render) の柔軟性のおかげで、他のアプリケーションに組み込むことができます。

React は一般的に起動時に単一のルート React コンポーネントを DOM にロードして使用されるものですが、`ReactDOM.render()` はボタンのような小さなものからアプリケーション全体に至るまで、独立した UI のパーツに対して複数回呼び出すこともできます。

実際、これはまさに React が Facebook で使用されている方法でもあります。これにより React でアプリケーションを少しずつ作成し、それらを既存のサーバー生成テンプレートやその他のクライアントサイドコードと組み合わせることができます。

> ### React で文字列ベースのレンダリングを置き換える {#replacing-string-based-rendering-with-react}

Reviewer: ```suggestion
### React で文字列ベースのレンダーを置き換える {#replacing-string-based-rendering-with-react}
```

> 古いウェブアプリケーションによくあるパターンは、DOM のまとまりを文字列として記述して、`$el.html(htmlString)` のような形で DOM に挿入することです。このような箇所は React の導入にぴったりです。文字列ベースのレンダリングを React コンポーネントに置き換えるだけで良いのです。

Reviewer: ```suggestion
古いウェブアプリケーションによくあるパターンは、DOM のまとまりを文字列として記述して、`$el.html(htmlString)` のような形で DOM に挿入することです。このような箇所は React の導入にぴったりです。文字列ベースのレンダーを React コンポーネントに置き換えるだけで良いのです。
```

);
```

> ここから始めて、コンポーネントにロジック部分を更に移植していくことや、より一般的な React のプラクティスを採用していくすることができます。例えば、コンポーネントでは同じコンポーネントが複数回レンダリングされる可能性があるので、ID に依存しないことがベストプラクティスです。かわりに React の [React event system](/docs/handling-events.html) を使用してクリックハンドラを React の `<button>` 要素に直接登録します：

Reviewer: ```suggestion
ここから始めて、コンポーネントにロジック部分を更に移植していくことや、より一般的な React のプラクティスを採用していくことができます。例えば、コンポーネントでは同じコンポーネントが複数回レンダーされる可能性があるので、ID に依存しないことがベストプラクティスです。かわりに React の [React event system](/docs/handling-events.html) を使用してクリックハンドラを React の `<button>` 要素に直接登録します：
```

[**Try it on CodePen**](http://codepen.io/gaearon/pen/RVKbvW?editors=1010)

> このような分離されたコンポーネントを好きなだけ持つことができ、`ReactDOM.render()` を使用して異なる DOM コンテナにそれらをレンダリングすることができます。アプリケーションを少しづつ React に変換していくにつれて、より大きなコンポーネントへとインテグレーションできるようになり、`ReactDOM.render()` の呼び出しを階層の上の方へ移動させていけるようになるでしょう。

Reviewer: ```suggestion
このような分離されたコンポーネントを好きなだけ持つことができ、`ReactDOM.render()` を使用して異なる DOM コンテナにそれらをレンダーすることができます。アプリケーションを少しずつ React に変換していくにつれて、より大きなコンポーネントへとインテグレーションできるようになり、`ReactDOM.render()` の呼び出しを階層の上の方へ移動させていけるようになるでしょう。
```

レンダリング→レンダー、少しづつ→少しずつ

> ### Backbone ViewにReactを組み込む {#embedding-react-in-a-backbone-view}

Reviewer: ```suggestion
### Backbone View に React を組み込む {#embedding-react-in-a-backbone-view}
```

[**Try it on CodePen**](http://codepen.io/gaearon/pen/gWgOYL?editors=0010)

`remove` メソッドで `ReactDOM.unmountComponentAtNode()` を呼び出して、コンポーネントツリーがデタッチされた際にイベントハンドラとコンポーネントツリーに関連付けられていたその他のリソースを React が解除することも重要です。

React *ツリー内*からコンポーネントが削除されるとクリーンアップは自動的に実行されますが、ツリー全体を手動で削除するため、このメソッドを呼び出す必要があります。

> ## Model層とのインテグレーション {#integrating-with-model-layers}

Reviewer: ```suggestion
## Model 層とのインテグレーション {#integrating-with-model-layers}
```

> 一般的には [React state](/docs/lifting-state-up.html)、[Flux](http://facebook.github.io/flux/)、もしくは [Redux](http://redux.js.org/) のような一方向のデータフローの使用が推奨されますが、React コンポーネントは他のフレームワークやライブラリのモデル層を利用することができます。

Reviewer: ```suggestion
一般的には [React の state](/docs/lifting-state-up.html)、[Flux](http://facebook.github.io/flux/)、もしくは [Redux](http://redux.js.org/) のような一方向のデータフローの使用が推奨されますが、React コンポーネントは他のフレームワークやライブラリのモデル層を利用することができます。
```

### React コンポーネントで Backbone Model を使用する {#using-backbone-models-in-react-components}

React コンポーネントから [Backbone](http://backbonejs.org/) のモデルとコレクションを利用する最もシンプルな方法は、様々な変更イベントを監視して手動で強制的に更新することです。

> モデルのレンダリングに責任をもつコンポーネントは `'change'` イベントを監視し、コレクションのレンダリングに責任をもつコンポーネントは `'add'` および `'remove'` イベントを監視します。どちらの場合も、[`this.forceUpdate()`](/docs/react-component.html#forceupdate) を呼び出して新しいデータでコンポーネントを再レンダリングします。

Reviewer: ```suggestion
モデルのレンダーに責任をもつコンポーネントは `'change'` イベントを監視し、コレクションのレンダーに責任をもつコンポーネントは `'add'` および `'remove'` イベントを監視します。どちらの場合も、[`this.forceUpdate()`](/docs/react-component.html#forceupdate) を呼び出して新しいデータでコンポーネントを再レンダリングします。
```

> 以下の例では、`List` コンポーネントは Backbone のコレクションをレンダリングします。個別の要素のレンダリングには `Item` コンポーネントを使用します。

Reviewer: ```suggestion
以下の例では、`List` コンポーネントは Backbone のコレクションをレンダーします。個別の要素のレンダーには `Item` コンポーネントを使用します。
```

[**Try it on CodePen**](http://codepen.io/gaearon/pen/GmrREm?editors=0010)

### Backbone Model からデータを抽出する {#extracting-data-from-backbone-models}

上記のアプローチでは React コンポーネントが Backbone のモデルとコレクションを認識することが必要です。後で別のデータ管理ソリューションに移行する予定がある場合は、Backbone とのインテグレーション箇所はできるだけコードの一部に集中させることをお勧めします。

> 1 つの解決法は、変更があるたびにモデルの属性をプレーンなデータとして抽出するようにし、そのロジックを一箇所だけで保持することです。以下のコードは Backbone モデルの全ての属性を state へと抽出し、そのデータをラップされるコンポーネントへと渡す [高階コンポーネント](/docs/higher-order-components.html) です。

Reviewer: ```suggestion
1 つの解決法は、変更があるたびにモデルの属性をプレーンなデータとして抽出するようにし、そのロジックを一箇所だけで保持することです。以下のコードは Backbone モデルの全ての属性を state へと抽出し、そのデータをラップされるコンポーネントへと渡す[高階コンポーネント](/docs/higher-order-components.html)です。
```

[**Try it on CodePen**](http://codepen.io/gaearon/pen/PmWwwa?editors=0010)

> この手法は Backbone だけに限ったものではありません。ライフサイクルメソッドの変更を購読し必要に応じてデータをローカルの React 状態にコピーすることで、任意のモデルライブラリで React を使用できます。

Reviewer: ```suggestion
この手法は Backbone だけに限ったものではありません。ライフサイクルメソッドで変更を購読し必要に応じてデータをローカルの React 状態にコピーすることで、任意のモデルライブラリで React を使用できます。
```

Author: 修正しました。ご確認ください。
